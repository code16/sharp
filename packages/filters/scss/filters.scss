
@import "components/FilterControl";
@import "components/FilterDateRange";
@import "components/FilterSelect";
@import "components/GlobalFilters";