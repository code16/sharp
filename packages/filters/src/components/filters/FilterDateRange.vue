<template>
    <div class="SharpFilterDateRange" :class="classes">
        <FilterControl :opened="opened" :label="label" :no-caret="noCaret" @click="handleClicked">
            <DateRange
                class="SharpFilterDateRange__field"
                :value="value"
                :display-format="displayFormat"
                :monday-first="mondayFirst"
                :clearable="!required"
                @input="handleInput"
                @focus="handlePickerFocused"
                @blur="handlePickerBlur"
                ref="range"
            />
        </FilterControl>
    </div>
</template>

<script>
    import { DateRange } from 'sharp-form';
    import FilterControl from '../FilterControl';

    export default {
        name: 'SharpFilterDateRange',

        components: {
            DateRange,
            FilterControl,
        },

        props: {
            value: {
                required: true,
            },
            required: Boolean,
            displayFormat: String,
            mondayFirst: Boolean,
            label: String,
        },

        data() {
            return {
                opened: false,
            }
        },

        computed: {
            empty() {
                return !this.value;
            },
            noCaret() {
                return !!this.value && !this.required;
            },
            classes() {
                return {
                    'SharpFilterDateRange--empty': this.empty,
                }
            },
        },

        methods: {
            handleClicked() {
                this.$refs.range.$refs.picker.focus();
            },
            handleInput(range) {
                this.$emit('input', range);
            },
            handlePickerFocused() {
                this.opened = true;
            },
            handlePickerBlur() {
                this.opened = false;
            },
        }
    }
</script>