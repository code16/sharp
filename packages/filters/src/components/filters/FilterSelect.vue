<template>
    <div class="SharpFilterSelect"
          :class="{
              'SharpFilterSelect--open':opened,
              'SharpFilterSelect--empty':empty,
              'SharpFilterSelect--multiple':multiple,
              'SharpFilterSelect--searchable':searchable
          }"
          tabindex="0"
    >
        <!-- dropdown & search input -->
        <Autocomplete
            class="SharpFilterSelect__select"
            :value="autocompleteValue"
            :local-values="values"
            :search-keys="searchKeys"
            :list-item-template="template"
            :placeholder="l('entity_list.filter.search_placeholder')"
            :multiple="multiple"
            :hide-selected="multiple"
            :allow-empty="!required"
            :preserve-search="false"
            :show-pointer="false"
            :searchable="searchable"
            no-result-item
            mode="local"
            ref="autocomplete"
            @multiselect-input="handleAutocompleteInput"
            @close="close"
        />

        <FilterControl :label="label" no-caret @click="handleClick">
            <!-- value text & tags -->
            <Select
                class="SharpFilterSelect__select"
                :value="value"
                :options="values"
                :multiple="multiple"
                :clearable="!required"
                :inline="false"
                placeholder=" "
                ref="select"
                @input="handleSelect"
            />
        </FilterControl>
    </div>
</template>

<script>
    import { Autocomplete, Select } from 'sharp-form'
    import { Localization } from 'sharp/mixins';
    import FilterControl from '../FilterControl';
    
    export default {
        name: 'SharpFilterSelect',
        mixins: [Localization],
        components: {
            Select,
            Autocomplete,
            FilterControl,
        },
        props: {
            label : {
                type: String,
                required: true
            },
            values: {
                type: Array,
                required: true
            },
            value: {
                type: [String, Number, Array],
            },
            multiple: Boolean,
            required: Boolean,
            searchable: Boolean,
            searchKeys: Array,
            template: String,
        },
        data() {
            return {
                opened: false,
                // dev only, check it in devtools to keep dropdown open
                debug: false,
            }
        },
        computed: {
            optionById() {
                return this.values.reduce((res, v)=> ({
                    ...res, [v.id]: v
                }), {});
            },
            empty() {
                return this.value == null || this.multiple && !this.value.length;
            },
            autocompleteValue() {
                return this.multiple ? (this.value||[]).map(value=>this.optionById[value]) : this.optionById[this.value];
            }
        },
        methods: {
            handleSelect(value) {
                this.$emit('input', value);
            },
            handleAutocompleteInput(value) {
                this.$emit('input', this.multiple ? value.map(v=>v.id) : (value||{}).id);
            },
            handleClick() {
                if(this.opened) {
                    this.close();
                } else {
                    this.open();
                }
            },
            open() {
                this.opened = true;
                this.$emit('open');
                this.$nextTick(this.showDropdown);
            },
            close() {
                if(this.debug) {
                    return;
                }
                this.opened = false;
                this.$emit('close');
                this.$nextTick(this.blur);
            },
            showDropdown() {
                let { autocomplete:{ $refs: { multiselect } } } = this.$refs;
                multiselect.activate();

                if(this.debug) {
                    this.unwatch && this.unwatch();
                    this.unwatch = multiselect.$watch('isOpen', function (isOpen) { if(!isOpen)this.isOpen = true; }, { sync: true });
                }
            },
            blur() {
                let { select:{ $refs: { multiselect } } } = this.$refs;
                multiselect.deactivate();
            },
        }
    }
</script>