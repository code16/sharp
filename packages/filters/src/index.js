import globalFilters from './store/global-filters';

export default function(Vue, { store }) {
    store.registerModule('global-filters', globalFilters);
}

export * from './components';
export { filters as filtersModule } from './store';