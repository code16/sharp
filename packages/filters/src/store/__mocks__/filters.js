
export default {
    namespaced: true,

    getters: {
        filters() {},
        values() {},
        getValuesFromQuery() {
            return jest.fn();
        },
        getQueryParams() {
            return jest.fn();
        },
        nextQuery() {
            return jest.fn();
        },
    }
}