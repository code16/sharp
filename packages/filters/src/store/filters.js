import Vue from 'vue';
import { parseRange, serializeRange } from "sharp";
import { getFiltersQueryParams, getFiltersValuesFromQuery, filterQueryKey } from '../util/query';

export const SET_FILTERS = 'SET_FILTERS';
export const SET_FILTER_VALUE = 'SET_FILTER_VALUE';

export default {
    namespaced: true,

    state: ()=>({
        filters: null,
        values: {},
    }),

    mutations: {
        [SET_FILTERS](state, filters) {
            state.filters = filters;
        },
        [SET_FILTER_VALUE](state, { key, value }) {
            Vue.set(state.values, key, value);
        },
    },

    getters: {
        value(state) {
            return key => state.values[key];
        },
        filters(state) {
            return state.filters ||Â []
        },
        values(state) {
            return state.values;
        },
        filter(state) {
            return key => (state.filters || []).find(filter => filter.key === key);
        },

        defaultValue() {
            return filter => (filter||{}).default;
        },
        isDateRange() {
            return filter => (filter||{}).type === 'daterange';
        },

        filterQueryKey() {
            return key => filterQueryKey(key);
        },
        getQueryParams(state, getters) {
            return values => {
                return getFiltersQueryParams(values, (value, key) =>
                    getters.serializeValue({ filter:getters.filter(key), value })
                );
            }
        },
        getValuesFromQuery() {
            return query => getFiltersValuesFromQuery(query);
        },
        resolveFilterValue(state, getters) {
            return ({ filter, value }) => {
                if(value == null) {
                    return getters.defaultValue(filter);
                }
                if(filter.multiple && !Array.isArray(value)) {
                    return [value];
                }
                if(getters.isDateRange(filter)) {
                    return parseRange(value);
                }
                return value;
            }
        },
        serializeValue(state, getters) {
            return ({ filter, value }) => {
                if(getters.isDateRange(filter)) {
                    return serializeRange(value);
                }
                return value;
            };
        },
        nextValues(state) {
            return ({ filter, value }) => {
                let base = state.values;
                if(filter.master) {
                    base = Object.keys(state.values).reduce((res, key) => ({ ...res, [key]:null }), {});
                }
                return { ...base, [filter.key]: value };
            };
        },
        nextQuery(state, getters) {
            return ({ filter, value }) => {
                return getters.getQueryParams(getters.nextValues({ filter, value }));
            }
        },
    },

    actions: {
        update({ commit, dispatch }, { filters, values }) {
            commit(SET_FILTERS, filters);

            return Promise.all(
                (filters || []).map(filter => {
                    dispatch('setFilterValue', {
                        filter,
                        value: (values || {})[filter.key]
                    })
                })
            );
        },
        setFilterValue({ commit, getters }, { filter, value }) {
            commit(SET_FILTER_VALUE, {
                key: filter.key,
                value: getters.resolveFilterValue({ filter, value })
            });
        },
    }
}