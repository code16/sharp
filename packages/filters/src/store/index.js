
export { default as filters } from './filters';