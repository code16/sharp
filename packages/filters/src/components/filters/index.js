import FilterDateRange from './FilterDateRange';
import FilterSelect from './FilterSelect';

export function filterByType(type) {
    if(type === 'select') {
        return FilterSelect;
    } else if(type === 'daterange') {
        return FilterDateRange;
    }
}