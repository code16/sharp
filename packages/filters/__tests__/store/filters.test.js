import filters from '../../src/store/filters';
import { SET_FILTERS, SET_FILTER_VALUE } from "../../src/store/filters";
import * as querystringUtils from "sharp/util/querystring";

describe('store filters', () => {
    beforeEach(() => {
        jest.restoreAllMocks()
    });

    test('state match snapshot', ()=>{
        expect(filters.state()).toMatchSnapshot();
    });

    describe('mutations', () => {
        test('SET_FILTERS', () => {
            const state = {};
            const testFilters = [];
            filters.mutations[SET_FILTERS](state, testFilters);

            expect(state.filters).toBe(testFilters);
        });

        test('SET_FILTER_VALUE', () => {
            const state = { values: {} };
            filters.mutations[SET_FILTER_VALUE](state, { key:'prop', value:'value' });

            expect(state.values.prop).toBe('value');
        });
    });

    describe('getters', () => {
        test('value', ()=>{
            const state = {
                values: {
                    prop: 'value'
                }
            };
            expect(filters.getters.value(state)('prop')).toBe('value');
        });

        test('filters', ()=>{
            expect(filters.getters.filters({ filter: null })).toEqual([]);
            expect(filters.getters.filters({ filters: [{}] })).toEqual([{}]);
        });

        test('filter', ()=>{
            expect(filters.getters.filter({ filters: null })('name')).toEqual(undefined);
            expect(
                filters.getters.filter({
                    filters: [{ key:'name' }],
                })('name')
            ).toEqual({ key:'name'});
        });

        test('defaultValue', ()=>{
            expect(filters.getters.defaultValue()(null)).toBeUndefined();
            expect(filters.getters.defaultValue()({ default:'default' })).toEqual('default');
        });

        test('isDateRange', ()=>{
            expect(filters.getters.isDateRange()(null)).toBe(false);
            expect(filters.getters.isDateRange()({ type:'daterange' })).toBe(true);
        });

        test('filterQueryKey', ()=>{
            expect(filters.getters.filterQueryKey()('key')).toBe('filter_key');
        });

        test('getQueryParams', ()=>{
            const state = {
            };
            const getters = {
                filterQueryKey: jest.fn(key => `filter_${key}`),
                filter: jest.fn(key => ({ key })),
                serializeValue: jest.fn(({ filter, value }) => value),
            };
            expect(filters.getters.getQueryParams(state, getters)({ })).toEqual({});

            expect(filters.getters.getQueryParams(state, getters)({
                type: 'aaa',
                name: 'bbb',
            })).toEqual({
                'filter_type': 'aaa', 'filter_name': 'bbb'
            });
            expect(getters.serializeValue).toHaveBeenCalledWith({ filter:{ key:'type' }, value:'aaa' });
            expect(getters.serializeValue).toHaveBeenCalledWith({ filter:{ key:'name' }, value:'bbb' });
        });

        test('getValuesFromQuery', ()=>{
            expect(filters.getters.getValuesFromQuery()(null)).toEqual({ });
            expect(filters.getters.getValuesFromQuery()({ filter_type: 'aaa', filter_name: 'bbb', custom: 'ccc' }))
                .toEqual({
                    type:'aaa', name: 'bbb'
                })
        });

        test('resolveFilterValue', ()=>{
            const state = {};
            const getters = {
                defaultValue: jest.fn(()=>'defaultValue'),
                isDateRange: jest.fn(()=>false),
            };
            const resolveFilterValue = (...args) => filters.getters.resolveFilterValue(state, getters)(...args);

            expect(resolveFilterValue({ filter: { key:'filter' }, value: undefined })).toEqual('defaultValue');
            expect(getters.defaultValue).toHaveBeenCalledWith({ key:'filter' });
            expect(resolveFilterValue({ filter: {}, value: null })).toEqual('defaultValue');

            expect(resolveFilterValue({
                filter:{ multiple: true }, value: 3
            })).toEqual([3]);

            expect(resolveFilterValue({
                filter: {}, value: 'test'
            })).toEqual('test');

            expect(resolveFilterValue({
                filter: { multiple: true }, value: [3]
            })).toEqual([3]);

            jest.spyOn(querystringUtils, 'parseRange')
                .mockImplementation(() => 'parsedRange');
            getters.isDateRange.mockReturnValue(true);
            expect(resolveFilterValue({
                filter: {}, value: '2019-06-21..2019-06-24',
            })).toEqual('parsedRange');
            expect(querystringUtils.parseRange)
                .toHaveBeenCalledWith('2019-06-21..2019-06-24');
        });

        test('serializeValue', ()=>{
            const state = {};
            const getters = {
                isDateRange: jest.fn(()=>false),
            };
            expect(filters.getters.serializeValue(state, getters)({
                filter: {},
                value: 'val'
            })).toEqual('val');

            jest.spyOn(querystringUtils, 'serializeRange')
                .mockImplementation(() => 'serializedRange');
            getters.isDateRange.mockReturnValue(true);
            expect(filters.getters.serializeValue(state, getters)({
                filter: {},
                value: {
                    start: 'start',
                    end: 'end',
                }
            })).toEqual('serializedRange');
            expect(querystringUtils.serializeRange).toHaveBeenCalledWith({ start:'start', end: 'end' });
        });

        test('nextValues', ()=>{
            const state = {
                values: {
                    type: 'aa'
                }
            };

            expect(filters.getters.nextValues(state)({ filter: { key:'filter' }, value: 1 }))
                .toEqual({
                    type: 'aa',
                    filter: 1
                });
            expect(filters.getters.nextValues(state)({ filter: { key:'filter', master: true }, value: 1 }))
                .toEqual({
                    type: null,
                    filter: 1
                });
        });

        test('nextQuery', ()=>{
            const getters = {
                getQueryParams: jest.fn(()=>'query params'),
                nextValues: jest.fn(()=>'next values')
            };

            expect(filters.getters.nextQuery(null, getters)({ filter:{ key:'filter' }, value:1 })).toEqual('query params');
            expect(getters.getQueryParams).toHaveBeenCalledWith('next values');
            expect(getters.nextValues).toHaveBeenCalledWith({ filter:{ key:'filter' }, value:1 });
        });
    });

    describe('actions', () => {
        test('update', async () => {
            const commit = jest.fn();
            const dispatch = jest.fn();
            const testFilters = [{ key:'prop1' }, { key:'prop2' }];
            const testValues = { prop1:'aaa', prop2:'bbb' };

            filters.actions.update({ commit, dispatch }, { filters: testFilters, values: testValues });

            expect(commit).toHaveBeenCalledWith(SET_FILTERS, testFilters);
            expect(dispatch).toHaveBeenCalledWith('setFilterValue', { filter: { key:'prop1' }, value: 'aaa' });
            expect(dispatch).toHaveBeenCalledWith('setFilterValue', { filter: { key:'prop2' }, value: 'bbb' });

            dispatch.mockClear();

            expect(() => {
                filters.actions.update({ commit, dispatch }, { filters: null, values: null });
            }).not.toThrow();

            expect(commit).toHaveBeenCalledWith(SET_FILTERS, null);
            expect(dispatch).not.toHaveBeenCalled();
        });

        test('setFilterValue', ()=>{
            const commit = jest.fn();
            const getters = {
                resolveFilterValue: jest.fn(()=>'resolvedValue')
            };
            const filter = { key:'filter' };
            const value = 'value';

            filters.actions.setFilterValue({ commit, getters }, { filter, value });

            expect(commit).toHaveBeenCalledWith(SET_FILTER_VALUE, { key: 'filter', value: 'resolvedValue' });
            expect(getters.resolveFilterValue).toHaveBeenCalledWith({ filter, value });
        });
    });
});