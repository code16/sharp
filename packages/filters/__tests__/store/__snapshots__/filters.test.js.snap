// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store filters state match snapshot 1`] = `
Object {
  "filters": null,
  "values": Object {},
}
`;
